### ISSUE TITLE
`pnpm add -g` fails on no global bin dir since version 7.20.0

### ISSUE THREAD

**Issue by FreekVR** [pnpm/pnpm#5883]
### pnpm version:
7.21.0, 7.22.0

### Code to reproduce the issue:
(In isolated/docker environment)
Docker image:
```
FROM node:16-bullseye-slim

# standard utilities we need
RUN apt-get update -qq && \
	apt-get install -y \
	    wget \
        curl \
        git \
        ca-certificates \
    --no-install-recommends && \
    rm -r /var/lib/apt/lists/*

# Install pnpm
RUN npm install -g pnpm@7
```

Run this code within a container using this image:
```
$ pnpm config set store-dir .pnpm-store
$ mkdir -p .pnpm-glob
$ mkdir -p .pnpm-bin
$ pnpm config set global-dir .pnpm-glob
$ pnpm config set global-bin-dir .pnpm-bin
$ export PATH=".pnpm-bin:$PATH"
$ pnpm --version
7.22.0
$ echo $PATH
.pnpm-bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
$ pnpm config list
//registry.npmjs.org/:_authToken=XXXX
global-bin-dir=.pnpm-bin
global-dir=.pnpm-glob
registry=https://registry.npmjs.org/
store-dir=.pnpm-store
user-agent=pnpm/7.22.0 npm/? node/v16.19.0 linux x64
$ pnpm add vercel -g
 ERR_PNPM_NO_GLOBAL_BIN_DIR  Unable to find the global bin directory
Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.
```

### Expected behavior:
Given that global-bin-dir option is set as per the instruction, it should not error.

### Actual behavior:
Returns an error, while it worked fine with the same code before this change was introduced; https://github.com/pnpm/pnpm/issues/5841.

### Additional information:

 - `node -v` prints:
 - 16.19.0
 - Windows, macOS, or Linux?:
 - Linux (node 16 docker image)

**Comment by FreekVR** [comment]
It seems this is actually occuring after 7.20.0, not 7.21.0, so may be related to this change;
https://github.com/pnpm/pnpm/pull/5829/files

**Comment by augustobritodev** [comment]
@FreekVR I tried 7.19.0 and i get the same error

**Comment by vaidotaslipskas** [comment]
tried with 7.23.0 Still same error on macOS M1

**Comment by gabygandrade** [comment]
+1 seeing the same issue with 7.26.0 on MacOS M1

**Comment by hxtree** [comment]
+same ubuntu dev container

**Comment by PlanCZero** [comment]
same issue with 7.26.0 on MacOS M2

**Comment by sethgw** [comment]
same

**Comment by leonardo-ornelas** [comment]
same

**Comment by maxkobzev** [comment]
actually can't update pnpm itself - `pnpm add -g pnpm` - the same error

**Comment by kenrick95** [comment]
Probably related to #4658 

@FreekVR did your issue occur only on these recent pnpm version or it happen to all pnpm 7 versions?

**Comment by FreekVR** [comment]
@kenrick95 No, older v7 versions are fine. I've pinned the version of pnpm in our CI/CD to `pnpm@7.19` which doesn't have this issue, only the more recent version (7.20.0 and 7.21.0 I last tested) have issues.

**Comment by elmeerr** [comment]
I was facing the same issue when trying to update from 7.16 to 7.29 and the issue was because I had node servers running...maybe that can help you.

> actually can't update pnpm itself - `pnpm add -g pnpm` - the same error

**Comment by zkochan** [comment]
@vickvasquez the pnpm home directory is not the store. It is confusing to name it the store. The pnpm store is located at these locations: https://pnpm.io/npmrc#store-dir

Also, this issue might be fixed as we shipped some fixes to the `pnpm config set` command.

**Comment by TechWatching** [comment]
@zkochan When running pnpm in a CI pipeline, what should we do to ensure  there is no "ERR_PNPM_NO_GLOBAL_BIN_DIR" error when executing the  `pnpm add -g ***` command  in a pipeline. I use it for instance to donwload the vercel CLI in order to deploy the package I have built.

**Comment by zkochan** [comment]
could you just use `pnpm dlx` instead? If you need to install a package globally just run it once, it is better to run `pnpm dlx`.

**Comment by TechWatching** [comment]
> could you just use `pnpm dlx` instead? If you need to install a package globally just run it once, it is better to run `pnpm dlx`.

Yes it does the job for my use case thanks 👍. Yet I can see some use cases where I would need to execute several commands from a CLI and `pnpm dlx` would not be the most appropriate.

**Comment by zkochan** [comment]
Yeah, `pnpm dlx` is currently reinstalling the package every time. We should make it so that repeat execution reuses the installation from the previous run (in some timeframe).

**Comment by shellscape** [comment]
I'm trying to use `pnpm add nx -g` within a github workflow and running into this after running `pnpm setup`:

```yml
      - name: Install Nx 😢
        env:
          SHELL: /bin/bash
        run: |
          pnpm setup
          source /home/runner/.bashrc
          pnpm add nx -g
```

Nx is already an inefficient runner (and I presently don't have a choice about using it on a project) and we're trying to avoid running it through another node process, running it directly which requires installing it globally.

**Comment by chokri** [comment]
I just added `pnpm setup --force`  and it worked for me.

**Comment by stephenhmarsh** [comment]
The solution here worked for me in the context of Alpine Linux in Docker: https://github.com/pnpm/pnpm/issues/784#issuecomment-1518582235

**Comment by lesar** [comment]
I'am on Fedora 38  
I have installed npm by nvm  
```
$ node -v
v20.9.0
$ npm -v
10.1.0
$ pnpm -v
8.10.5
```
if I make:  
`$ pnpm setup`  

then   

```
$ pnpn add -g pnpm
Nothing to stop. No server is running for the store at /home/leonardo/.local/share/pnpm/store/v3
Already up to date
Progress: resolved 1, reused 0, downloaded 0, added 0, done
Done in 2.5s
```  

work fine (no error)  

I have try to follow nvm pnpm executable on 
`/home/leonardo/.nvm/versions/node/v20.9.0/lib/node_modules/corepack/shims/pnpm`  

 but after a while my eyes crossed.  
typing this:  
```
$ /home/leonardo/.nvm/versions/node/v20.9.0/lib/node_modules/corepack/shims/pnpm -v
8.10.5
$ ~/.local/share/pnpm/pnpm -v
8.10.5
```
show the same version ...
but I didn't understand much about it.  
I'm not sure they are the same executable but at the last have the same version 

My question is:
Anyone know if in my use case  
(npm by nvm and pnpm by npm)  
if there will be any future ***hidden problems*** to solve  
`pnpm add -g pnpm`  
by doing  
`pnpm setup`  
like I have doing!?

best regards,
Leonardo

**Comment by kale1d0code** [comment]
I've been having this  problem when trying to install global packages during postStartCommand in a devcontainer.

no matter what I do, sourcing the .bashrc file makes no difference.
it works fine after the container has started, although I kind of needed my global packages installed before that point.

**Comment by ZowWeb** [comment]
Run these 2 simple commands
1. `pnpm setup`
2. `source ~/.zshrc`

it should now work!

**Comment by Eprince-hub** [comment]
I also witnessed this error, and nothing was helping. After running `pnpm setup` multiple times without any change, I found that  running `pnpm config set global-bin-dir <path to pnpm>`  worked

I am using pnpm version 9.15.4

**Comment by NymuGit** [comment]
Finally, i found some unique workaround...

If pnpm are installed either from winget or powershell script, run `pnpm setup` and `pnpm add -g some-package` will always throw global-bin-dir error.

Check pnpm path using:
```
$env:Path
```

Find pnpm path (usually like this):
```
...
C:\Users\<username>\AppData\Local\pnpm
...
```
then, open Explorer (WIN + E) and go to desired location, create empty folder named "global"

(or if you using powershell on `C:\Users\<username>` path):
```
mkdir AppData\Local\pnpm\global
```

Relaunch Powershell and try install package using pnpm (except pnpm, because it throw an error that guide you using pnpm self-update):
```
pnpm add -g some-package
```

That's it. I hope it worked for you :)
`>>  n7mu`

note: This is for Windows 11. Currently using 2H24 build. This workaround should work too for earlier versions of Windows.

**Comment by manosbatsis** [comment]
Thanks @Eprince-hub,  your fix saved my CI:

```
FROM node:22.22-bookworm-slim AS base
# ...
ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable
RUN pnpm config set global-bin-dir $PNPM_HOME # <-- this fixed the build
# ...

FROM base AS pruner
RUN apt update && rm -rf /var/lib/apt/lists/*
WORKDIR /app
RUN pnpm add -g turbo # <-- was failing before the fix
```

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm add -g` installs, and when the bin directory is not on PATH the error names the directory and says to run `pnpm setup` (verified while checking 4962).

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
